### PR TITLE
Add shopinvader_locomotive_impersonate

### PR DESCRIPTION
**Pull request by sebastienbeau**

### CODE REVIEW

**Comment by simahawk** [comment]
Requires https://github.com/shopinvader/locomotive-shopinvader/pull/77 I guess...

**Comment by sebastienbeau** [comment]
Yes and it have been merge and release
Last docker image here : https://github.com/orgs/shopinvader/packages?repo_name=docker-locomotive-shopinvader

**Review by ivantodorovich — APPROVED** [review]

File: shopinvader_locomotive_impersonate/models/shopinvader_partner.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+    def _get_from_token(self, email, token):
+        now = fields.Datetime.now()
+        partner = self.search(
+            [
+                ("impersonate_token", "=", token),
+                ("email", "=", email),
+                ("datetime_expire_impersonate_token", ">", now),
+            ]
```
Comment (simahawk) [inline shopinvader_locomotive_impersonate/models/shopinvader_partner.py]:
```suggestion
            ], limit=1
```

File: shopinvader_locomotive_impersonate/services/customer.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    _inherit = "shopinvader.customer.service"
+
+    @restapi.method(
+        routes=[(["/impersonate"], "GET")],
+        input_param=restapi.CerberusValidator("_validator_impersonate"),
+    )
+    def impersonate(self, email=None, token=None):
+        shop_partner = self.env["shopinvader.partner"]._get_from_token(email, token)
+        if shop_partner:
+            customer = self._to_customer_info(shop_partner.record_id)
```
Comment (simahawk) [inline shopinvader_locomotive_impersonate/services/customer.py:27]: shall we log something specific here?  Eg:
```
_logger.info("Shopinvader customer impersonate: %s", email)
```

**Review by simahawk — APPROVED** [review]
LG overall, small remarks

File: shopinvader_locomotive_impersonate/models/shopinvader_partner.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import uuid
+
+from odoo import _, fields, models
+from odoo.exceptions import AccessError
+
+
+class ShopinvaderPartner(models.Model):
+    _inherit = "shopinvader.partner"
+
+    impersonate_token = fields.Char()
```
Comment (ivantodorovich) [inline shopinvader_locomotive_impersonate/models/shopinvader_partner.py]:
```suggestion
    impersonate_token = fields.Char(index=True)
```

File: shopinvader_locomotive_impersonate/services/customer.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            return {
+                "store_cache": {"customer": customer},
+                "force_authenticate_customer": shop_partner.email,
+                "redirect_to": shop_partner.backend_id.location,
+            }
+        else:
+            raise UserError(_("Invalid impersonate token"))
```
Comment (ivantodorovich) [inline shopinvader_locomotive_impersonate/services/customer.py]:
```suggestion
            raise AccessError(_("Invalid impersonate token"))
```
maybe it's more accurate?

**Comment by sebastienbeau** [comment]
@simahawk @ivantodorovich I have done the requested changes, it's ok for you ?

File: shopinvader_locomotive_impersonate/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Shopinvader Locomotive Impersonate",
+    "summary": "Allow to log with any shopinvader partner directly from odoo",
+    "version": "14.0.1.0.0",
+    "development_status": "Alpha",
+    "category": "Shopinvader",
+    "website": "https://github.com/shopinvader/odoo-shopinvader",
+    "author": " Akretion",
```
Comment (ivantodorovich) [inline shopinvader_locomotive_impersonate/__manifest__.py]:
```suggestion
    "author": "Akretion",
```

**Review by ivantodorovich — APPROVED** [review]
LGTM

**Pull request merged**